---
title: "Untitled"
author: "Joke Deschildre"
date: "10 februari 2020"
output: html_document
version: "R version 4.0.3 (2020-10-10) -- Bunny-Wunnies Freak Out"
---

```{r}
library(data.table)
library(stringr)
```
# Sample info 
Table with the general sample info 20Q4 v2 downloaded 10/03/2021 from the DepMap website. 
```{r}
sample_info <- read.csv('20Q4 v2/20Q4_v2_sample_info.csv', header=TRUE, sep=",", fill=TRUE)
sample_info
```


# Expression data
The RNAseq_reads file 20Q4 v2 downloaded 10/03/2021 from the DepMap website. 
```{r}
RNA_seq_table <- fread("20Q4 v2/20Q4_v2_CCLE_RNAseq_reads.csv", data.table = F)
#head(RNA_seq_table)
print(length(RNA_seq_table[,1]))
RNA_seq_cell_lines <- lapply(RNA_seq_table[,1], function(x) sample_info[match(x, sample_info[,1]),4])
RNA_seq_cell_lines <- unlist(RNA_seq_cell_lines)
print(length(RNA_seq_cell_lines)) # all cell lines are found in the sample info file
print(dim(sample_info))
sample_info_expr_data <- sample_info[ sample_info$CCLE_Name %in% RNA_seq_cell_lines , ] #make new sample info table with only cell lines also present in expression data considered 
print(dim(sample_info_expr_data)) # MANY CELL LINES NOT PRESENT IN EXPRESSION CELL LINES!!

```

The sample info for certain tissue types can be printed seperately. 
```{r}
sample_info_expr_data[grep("HAEMATOPOIETIC_AND_LYMPHOID_TISSUE", sample_info_expr_data$CCLE_Name),]
```
#Overview of cell line types 

Overview table of the different tissue types for which RNA data is present. 
```{r}
tissue_sample_info_expr_data <- str_remove(sample_info_expr_data$CCLE_Name, "^[a-zA-Z0-9]+_") #remove the cell line code/name to be left with the tissue type
print(unique(tissue_sample_info_expr_data))
data.frame(table(tissue_sample_info_expr_data)) #the found totals per tissue type do (mostly) agree to the to totals stated in Jens' file
```
Overview table of the different diseases/cancers for which RNA data is present. 
```{r}
data.frame(table(sample_info_expr_data$primary_disease)) #!! does not completely agree to tissue types!!! 
```
#Outlier cell lines 
Table of outlier cell lines (see GitHub page of the lab, Jens' project): Jens investigated which cell lines do not represent well the according TCGA cancer type. He did so based on 1) the Yu paper "Comprehensive transcriptomic analysis of cell lines as models of primary tumors across 22 tumor types" 2) clustering. 
```{r}
outlier_table <- fread("C:/Users/joked/OneDrive - ugentbe/PhD Joke shared/paper sample specific benchmarking/cell lines decision/Jens 20Q4 data/BlackList_sample_outlier1.csv", data.table=F)
outlier_table
```
#Overlap tables 
Funtion to make an overlap table: which cell lines 1) are outliers 2) have certain data types available + extra annotation
```{r} 
#TO ADD: SAMPLE INFO
overlap_table_fun <- function(sample_info_expr_data, data_list, tissue_type, remove_outlier_cell_lines = TRUE, outlier_table){
  # I work with sample_info_expr_data because I am only interested in cell lines present in the mRNA data (also Jens only took into account mRNA cell lines)
  disease_cell_lines_CCLE_name <- sample_info_expr_data$CCLE_Name[grep(tissue_type, sample_info_expr_data$CCLE_Name)]
  overlap_table <- data.frame(matrix(NA, nrow = length(disease_cell_lines_CCLE_name) , ncol = length(data_list)+ remove_outlier_cell_lines*2+3))
  rownames(overlap_table) <- disease_cell_lines_CCLE_name
  colnames(overlap_table) <- c(names(data_list), c("sample collection site", "primary or metastasis", "primary disease"))
  
  if (remove_outlier_cell_lines==T){ #add columns with information about the outlier cell lines based on Jens' analysis
    colnames(overlap_table) <- c('non_Yu_outliers','non_hclust_outliers', c(names(data_list), c("sample collection site", "primary or metastasis", "primary disease")))
    overlap_table[,1] <- outlier_table[match(rownames(overlap_table), outlier_table[,1]),2]
    overlap_table[is.na(overlap_table[,1]), 1] <- 0
    overlap_table[, 1] <- 1 - overlap_table[,1] # make zeros ones and ones zeros 
    overlap_table[,2] <- outlier_table[match(rownames(overlap_table), outlier_table[,1]),3]
    overlap_table[is.na(overlap_table[,2]), 2] <- 0
    overlap_table[, 2] <- 1 - overlap_table[,2]  # make zeros ones and ones zeros
  }
  for (i in (1+remove_outlier_cell_lines*2):(length(data_list)+remove_outlier_cell_lines*2)){ # add columns with information about the different data types 
    overlap_table[,i] <- as.numeric(rownames(overlap_table) %in% data_list[[i-remove_outlier_cell_lines*2]])
  }
  indices_cell_lines_sample_info <- match(rownames(overlap_table), sample_info_expr_data$CCLE_Name) # add columns with extra annotation from the sample_info file 
  overlap_table[,(length(data_list)+remove_outlier_cell_lines*2)+1] <- sample_info_expr_data$sample_collection_site[indices_cell_lines_sample_info]
  overlap_table[,(length(data_list)+remove_outlier_cell_lines*2)+2] <- sample_info_expr_data$primary_or_metastasis[indices_cell_lines_sample_info]
  overlap_table[,(length(data_list)+remove_outlier_cell_lines*2)+3] <- sample_info_expr_data$primary_disease[indices_cell_lines_sample_info]
  
  return(overlap_table)
}
```


Tryout of the function overlap_table_fun for the breast tissue type 
```{r}
overlap_table_fun(sample_info_expr_data, data.frame(RNA_seq=RNA_seq_cell_lines), "BREAST", remove_outlier_cell_lines = T, outlier_table = outlier_table)
```

# CRISPR knockdown data
the  Ach_gene_dependency file 20Q4 v2 from the DEPMAP website (CRISPR knockdown data) downloaded 10/03/2021. 
"This Achilles dataset contains the results of genome-scale CRISPR knockout screens for 18,333 genes in 739 cell lines"
```{r}
CRISPR_table <- fread("20Q4 v2/20Q4_v2_Achilles_gene_dependency.csv", data.table = F)
#head(CRISPR_table)
CRISPR_cell_lines <- lapply(CRISPR_table[,1], function(x) sample_info_expr_data[match(x, sample_info_expr_data[,1]),4])
CRISPR_cell_lines <- unlist(CRISPR_cell_lines)
length(CRISPR_cell_lines)
```

different achilles files contain the same cell lines: 
```{r}
#CRISPR_table_2 <- fread("20Q4_v2_Achilles_gene_effect.csv", data.table = F)
#head(CRISPR_table_2)
#CRISPR_cell_lines_2 <- lapply(CRISPR_table_2[,1], function(x) sample_info[match(x, sample_info[,1]),4])
#CRISPR_cell_lines_2 <- unlist(CRISPR_cell_lines_2)
#length(CRISPR_cell_lines)
#length(CRISPR_cell_lines_2)
#sum(CRISPR_cell_lines %in% CRISPR_cell_lines_2)
#sum(CRISPR_cell_lines_2 %in% CRISPR_cell_lines)
```

# Mutation data
the CCLE_mutations file 20Q4 v2 from the DEPMAP website downloaded 10/03/2021.
```{r}
mutation_table <- fread("20Q4 v2/20Q4_v2_CCLE_mutations.csv", data.table = F)
#head(mutation_table)
mutation_cell_lines <- lapply(unique(mutation_table$DepMap_ID), function(x) sample_info_expr_data[match(x, sample_info_expr_data[,1]),4])
mutation_cell_lines <- unlist(mutation_cell_lines)
length(mutation_cell_lines)
```


# RRBS data (methylation)
the RRBS file (CCLE_RRBS_tss_CpG_clusters_20181022.txt and annotation file CCLE_Cell_lines_annotations_20181226.txt ) from the CCLE website downloaded 24/11/2020.
```{r}
RRBS_table <- fread("20Q4 v2/CCLE_RRBS_tss_CpG_clusters_20181022.txt", data.table = F)
RRBS_cell_lines <- colnames(RRBS_table)[-c(1,2,3)]
RRBS_cell_lines <- lapply(RRBS_cell_lines, function(x) sample_info_expr_data[match(x, sample_info_expr_data[,4]),4])

RRBS_cell_lines <- unlist(RRBS_cell_lines)
length(RRBS_cell_lines)
```

Other RRBS files contain the same cell lines 
```{r}
#RRBS_table_2 <- fread("20Q4 v2/CCLE_RRBS_TSS1kb_20181022.txt", data.table = F)
#RRBS_cell_lines_2 <- colnames(RRBS_table_2)[-c(1,2,3)]
#length(RRBS_cell_lines_2)
#sum(RRBS_cell_lines==RRBS_cell_lines_2)
```


#Copy number data 
The copy number data 20Q4 v2 from the DepMap website, downloaded 10/03/2020
```{r}
CN_table <- fread("20Q4 v2/20Q4_v2_CCLE_gene_cn.csv", data.table = F)
#head(CN_table)
CN_cell_lines <- lapply(CN_table[,1], function(x) sample_info_expr_data[match(x, sample_info_expr_data[,1]),4])
CN_cell_lines <- unlist(CN_cell_lines)
length(CN_cell_lines)
```
#Protein Data
Protein data from DepMap, from Gygi lab, downloaded 24/11/2020
```{r}
protein_table <- fread("20Q4 v2/Gygi_Proteomics_Table_S2_Protein_Quant_Normalized.csv", data.table = F)
#head(protein_table)
protein_cell_lines <- str_remove(colnames(protein_table)[49:426], "_TenPx[0-9]{2}")
protein_cell_lines <- lapply(protein_cell_lines, function(x) sample_info_expr_data[match(x, sample_info_expr_data[,4]),4])
protein_cell_lines <- unlist(protein_cell_lines)
#protein_cell_lines
length(protein_cell_lines)
```

#drug response data
GDSC as found on their site https://www.cancerrxgene.org/ downloaded 24/11/2020
```{r}
GDSC_table_2 <- fread("20Q4 v2/GDSC2_fitted_dose_response_25Feb20.csv", data.table = F)
GDSC_table_1 <- fread("20Q4 v2/GDSC1_fitted_dose_response_25Feb20.csv", data.table = F)
#head(GDSC_table)
GDSC_cell_lines_2 <- unique(GDSC_table_2$CELL_LINE_NAME)
GDSC_cell_lines_1 <- unique(GDSC_table_1$CELL_LINE_NAME)
GDSC_cell_lines <- unique(GDSC_cell_lines_1, GDSC_cell_lines_2)
GDSC_cell_lines <- str_remove_all(GDSC_cell_lines, "[-\\(\\)\\[\\]]") # -()[] characters are present in GDSC files but not DepMap files
GDSC_cell_lines <- toupper(GDSC_cell_lines) # lowercase letters are present int GDSC files but not DepMap files 
GDSC_cell_lines <- lapply(GDSC_cell_lines, function(x) sample_info_expr_data[match(x, sample_info_expr_data[,3]),4])
GDSC_cell_lines <- unlist(GDSC_cell_lines)
GDSC_cell_lines <- as.character(GDSC_cell_lines[!is.na(GDSC_cell_lines)])
length(GDSC_cell_lines)
```
GDSC1 and 2 as found on the DepMap website --> 2 more cell lines found, downloaded 15/12/2020
```{r}
sanger_table <- fread("20Q4 v2/GDSC1_GDSC2_sanger-dose-response.csv", data.table = F)
sanger_cell_lines <- unique(sanger_table$ARXSPAN_ID)
sanger_cell_lines <- lapply(sanger_cell_lines, function(x) sample_info_expr_data[match(x, sample_info_expr_data[,1]),4])
sanger_cell_lines <- unlist(sanger_cell_lines)
sanger_cell_lines <- as.character(sanger_cell_lines[!is.na(sanger_cell_lines)])
length(sanger_cell_lines)
```
Which cell lines are unique for the sanger_table and the GDSC_table
```{r}
sanger_cell_lines[!(sanger_cell_lines %in% GDSC_cell_lines)]
GDSC_cell_lines[!(GDSC_cell_lines %in% sanger_cell_lines)]
```

PRISM from DepMap website downloaded 27/11/2020
```{r}
PRISM_table <- fread("20Q4 v2/drug_sens_secondary-screen-dose-response-curve-parameters.csv", data.table = F)
PRISM_cell_lines <- unique(PRISM_table$ccle_name)
PRISM_cell_lines <- lapply(PRISM_cell_lines, function(x) sample_info_expr_data[match(x, sample_info_expr_data[,4]),4])
PRISM_cell_lines <- unlist(PRISM_cell_lines)
length(PRISM_cell_lines)
```
CTRP drug sensitivity data downloaded 27/11/2020 https://portals.broadinstitute.org/ctrp/
```{r}
CTRP_table <- fread("20Q4 v2/CTRP_v20.data.curves_post_qc.txt", data.table=F)
CTRP_experiment_annotation <- fread("20Q4 v2/CTRP_v20.meta.per_experiment.txt", data.table=F)
CTRP_cell_line_annotation <- fread("20Q4 v2/CTRP_v20.meta.per_cell_line.txt", data.table=F)

experiments_in_table <- unique(CTRP_table$experiment_id)
cell_lines_id_in_table <- lapply(experiments_in_table, function(x) CTRP_experiment_annotation[match(x, CTRP_experiment_annotation[,1]),9])
cell_lines_stripped_in_table <- lapply(cell_lines_id_in_table, function(x) CTRP_cell_line_annotation[match(x, CTRP_cell_line_annotation[,1]),2])
CTRP_cell_lines <- lapply(cell_lines_stripped_in_table, function(x) sample_info_expr_data[match(x, sample_info_expr_data[,3]),4])
CTRP_cell_lines <- unlist(CTRP_cell_lines)
length(CTRP_cell_lines)
CTRP_cell_lines <- as.character(CTRP_cell_lines[!is.na(CTRP_cell_lines)])
length(CTRP_cell_lines)
```



# Overlap table for different tissue types 
```{r}
data_list <- list(RNA_seq_cell_lines, mutation_cell_lines, CN_cell_lines, protein_cell_lines, RRBS_cell_lines, CRISPR_cell_lines, sanger_cell_lines, GDSC_cell_lines, PRISM_cell_lines, CTRP_cell_lines)
names(data_list) <- c('RNA', 'mutation', 'CN', 'protein', 'RRBS', 'CRISPR', 'GDSC drug sensitivity(DepMap)', 'GDSC drug sensitivity(GDSC)', 'PRISM drug sensitivity', 'CTRP drug sensitivity')
#LUNG
overlap_table <- overlap_table_fun(sample_info_expr_data, data_list, "LUNG", remove_outlier_cell_lines = TRUE, outlier_table)
write.csv(overlap_table, "overlap_table_LUNG.csv")
#HAEMATOPOIETIC_AND_LYMPHOID_TISSUE
overlap_table <- overlap_table_fun(sample_info_expr_data, data_list, "HAEMATOPOIETIC_AND_LYMPHOID_TISSUE", remove_outlier_cell_lines = TRUE, outlier_table)
write.csv(overlap_table, "overlap_table_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE.csv")
#breast
overlap_table <- overlap_table_fun(sample_info_expr_data, data_list, "BREAST", remove_outlier_cell_lines = TRUE, outlier_table)
write.csv(overlap_table, "overlap_table_BREAST.csv")
#LARGE_INTESTINE
overlap_table <- overlap_table_fun(sample_info_expr_data, data_list, "LARGE_INTESTINE", remove_outlier_cell_lines = TRUE, outlier_table)
write.csv(overlap_table, "overlap_table_LARGE_INTESTINE.csv")
#SKIN
overlap_table <- overlap_table_fun(sample_info_expr_data, data_list, "SKIN", remove_outlier_cell_lines = TRUE, outlier_table)
write.csv(overlap_table, "overlap_table_SKIN.csv")
#PANCREAS
overlap_table <- overlap_table_fun(sample_info_expr_data, data_list, "PANCREAS", remove_outlier_cell_lines = TRUE, outlier_table)
write.csv(overlap_table, "overlap_PANCREAS.csv")
#FIBROBLAST
overlap_table <- overlap_table_fun(sample_info_expr_data, data_list, "FIBROBLAST", remove_outlier_cell_lines = TRUE, outlier_table)
write.csv(overlap_table, "overlap_table_FIBROBLAST.csv")
#CENTRAL_NERVOUS_SYSTEM
overlap_table <- overlap_table_fun(sample_info_expr_data, data_list, "CENTRAL_NERVOUS_SYSTEM", remove_outlier_cell_lines = TRUE, outlier_table)
write.csv(overlap_table, "overlap_table_CENTRAL_NERVOUS_SYSTEM.csv")
#OVARY
overlap_table <- overlap_table_fun(sample_info_expr_data, data_list, "OVARY", remove_outlier_cell_lines = TRUE, outlier_table)
write.csv(overlap_table, "overlap_table_OVARY.csv")
#UPPER_AERODIGESTIVE_TRACT
overlap_table <- overlap_table_fun(sample_info_expr_data, data_list, "UPPER_AERODIGESTIVE_TRACT", remove_outlier_cell_lines = TRUE, outlier_table)
write.csv(overlap_table, "overlap_table_UPPER_AERODIGESTIVE_TRACT.csv")
#STOMACH
overlap_table <- overlap_table_fun(sample_info_expr_data, data_list, "STOMACH", remove_outlier_cell_lines = TRUE, outlier_table)
write.csv(overlap_table, "overlap_table_STOMACH.csv")
#ENDOMETRIUM
overlap_table <- overlap_table_fun(sample_info_expr_data, data_list, "ENDOMETRIUM", remove_outlier_cell_lines = TRUE, outlier_table)
write.csv(overlap_table, "overlap_table_ENDOMETRIUM.csv")
#OESOPHAGUS
overlap_table <- overlap_table_fun(sample_info_expr_data, data_list, "OESOPHAGUS", remove_outlier_cell_lines = TRUE, outlier_table)
write.csv(overlap_table, "overlap_table_OESOPHAGUS.csv")
#UPPER_AERODIGESTIVE_TRACT
overlap_table <- overlap_table_fun(sample_info_expr_data, data_list, "UPPER_AERODIGESTIVE_TRACT", remove_outlier_cell_lines = TRUE, outlier_table)
write.csv(overlap_table, "overlap_table_UPPER_AERODIGESTIVE_TRACT.csv")
#SOFT_TISSUE
overlap_table <- overlap_table_fun(sample_info_expr_data, data_list, "SOFT_TISSUE", remove_outlier_cell_lines = TRUE, outlier_table)
write.csv(overlap_table, "overlap_table_SOFT_TISSUE.csv")
#bone
overlap_table <- overlap_table_fun(sample_info_expr_data, data_list, "BONE", remove_outlier_cell_lines = TRUE, outlier_table)
write.csv(overlap_table, "overlap_table_BONE.csv")
#KIDNEY
overlap_table <- overlap_table_fun(sample_info_expr_data, data_list, "KIDNEY", remove_outlier_cell_lines = TRUE, outlier_table)
write.csv(overlap_table, "overlap_table_KIDNEY.csv")
#BILIARY_TRACT
overlap_table <- overlap_table_fun(sample_info_expr_data, data_list, "BILIARY_TRACT", remove_outlier_cell_lines = TRUE, outlier_table)
write.csv(overlap_table, "overlap_table_BILIARY_TRACT.csv")


```

After analysing the resulting overlap tables per tissue type, cell lines can be removed based on available data & characteristics.

For the central nervous system tissue cell lines all hclust and Yu outliers are removed, all metastasis samples and all samples that do not mention brain cancer as the primary disease. 
The final expression data we will used for furthere analysis is stored in a csv file.
```{r}
brain_overlap_table <- fread("overlap_table_CENTRAL_NERVOUS_SYSTEM.csv")
print(dim(brain_overlap_table))
brain_overlap_table <- brain_overlap_table[brain_overlap_table$non_Yu_outliers==1] #remove Yu outliers
print(dim(brain_overlap_table))
brain_overlap_table <- brain_overlap_table[brain_overlap_table$non_hclust_outliers==1] # remove hclust outliers
print(dim(brain_overlap_table))
brain_overlap_table <- brain_overlap_table[brain_overlap_table$`primary or metastasis`!="Metastasis"] # remove metastasis samples
print(dim(brain_overlap_table))
brain_overlap_table <- brain_overlap_table[brain_overlap_table$`primary disease`=="Brain Cancer"] #remove all non brain cancer samples
print(dim(brain_overlap_table))
brain_overlap_table
remaining_brain_cell_lines <- brain_overlap_table[,1]
remaining_brain_cell_lines_id <- sample_info_expr_data$DepMap_ID[match(unlist(remaining_brain_cell_lines), sample_info_expr_data$CCLE_Name)]
remaining_brain_RNA_seq_table <- RNA_seq_table[match(remaining_brain_cell_lines_id, RNA_seq_table[,1]),]
remaining_brain_RNA_seq_table[,1]
fwrite(remaining_brain_RNA_seq_table, "selection_brain_expression_data.csv") #TAKES LONG!!!!
```
For the lung tissue cell lines all hclust and Yu outliers are removed, all metastasis samples and all samples that do not mention lung cancer as the primary disease. 
The final expression data we will used for furthere analysis is stored in a csv file.
```{r}
lung_overlap_table <- fread("overlap_table_LUNG.csv")
print(dim(lung_overlap_table))
lung_overlap_table <- lung_overlap_table[lung_overlap_table$non_Yu_outliers==1] #remove Yu outliers
print(dim(lung_overlap_table))
lung_overlap_table <- lung_overlap_table[lung_overlap_table$non_hclust_outliers==1] # remove hclust outliers
print(dim(lung_overlap_table))
lung_overlap_table <- lung_overlap_table[lung_overlap_table$`primary or metastasis`!="Metastasis"] # remove metastasis samples
print(dim(lung_overlap_table))
lung_overlap_table <- lung_overlap_table[lung_overlap_table$`primary disease`=="Lung Cancer"] #remove all non lung cancer samples
print(dim(lung_overlap_table))
lung_overlap_table
remaining_lung_cell_lines <- lung_overlap_table[,1]
remaining_lung_cell_lines_id <- sample_info_expr_data$DepMap_ID[match(unlist(remaining_lung_cell_lines), sample_info_expr_data$CCLE_Name)]
remaining_lung_RNA_seq_table <- RNA_seq_table[match(remaining_lung_cell_lines_id, RNA_seq_table[,1]),]
remaining_lung_RNA_seq_table[,1]
fwrite(remaining_lung_RNA_seq_table, "selection_lung_expression_data.csv") #TAKES LONG!!!!
```

